### PR TITLE
KEP-1635 More efficient crypto

### PR DESCRIPTION
**Pull request by MontyThibault**
- Switched to native nodejs crypto for performance speedups.
- A browser fallback is used (https://github.com/crypto-browserify/crypto-browserify) 
- The first commit was tested by Nic and yielded significant performance improvements.
- The second commit was tested to be functional in Chrome.

### CODE REVIEW

File: src/swarmClient/main.js [hunk 1]
```diff
@@ -39,16 +36,6 @@ module.exports = {
         onclose = (onclose && once(onclose)) || (() => {});
 
 
-
-        if(public_pem) {
-            // throws an error if key is malformed
```
Comment (nickroeker) [inline src/swarmClient/main.js]: We're losing this up-front verification, correct? Seems like a regression in startup behaviour.

File: src/swarmClient/main.js [hunk 2]
```diff
@@ -39,16 +36,6 @@ module.exports = {
...
-            import_public_key_from_base64(public_pem);
-        }
-
-        public_pem = public_pem || pub_from_priv(private_pem);
```
Comment (nickroeker) [inline src/swarmClient/main.js]: If I'm reading correctly, we used to generate a public key if only the private key was given. Now, we do not. Is this the case? This may also be a regression in startup behaviour.

File: src/test/main.test.js [hunk 3]
```diff
@@ -39,8 +38,16 @@ const logDetailed = true;
 
 
 // these mirror the keys in scripts/run-swarms.rb
-const master_pub_key = "MFYwEAYHKoZIzj0CAQYFK4EEAAoDQgAEE/Yeq9sYdyeou+TnNEJjMnuntrzqcFIfIHd49LW461d55TY4hVX66ZXXGvAWRqMVMeELtYuKGYU44bPaxTb1ig==";
-const master_priv_key = "MHQCAQEEIEOd7E9zSxgJjtpGzK/gHl0vVSOZ2iF3TY50InD67BnHoAcGBSuBBAAKoUQDQgAEE/Yeq9sYdyeou+TnNEJjMnuntrzqcFIfIHd49LW461d55TY4hVX66ZXXGvAWRqMVMeELtYuKGYU44bPaxTb1ig==";
+const master_pub_key = "-----BEGIN PUBLIC KEY-----\n\
```
Comment (nickroeker) [inline src/test/main.test.js]: This is a change in contract to our users. Do we gain any benefit from requiring the header/footer? If not, I'd rather keep the user expectation the same.

File: src/test/main.test.js [hunk 4]
```diff
@@ -117,22 +98,6 @@ it('version', () => {
 });
 
 
-it('rejects operations on a non-extant database', async () => {
```
Comment (nickroeker) [inline src/test/main.test.js]: Why was this removed?

File: src/test/main.test.js [hunk 5]
```diff
@@ -316,7 +281,7 @@ describe('api', function() {
     
 
 
-    it('writers', async () => {
+    it.skip('writers', async () => {
```
Comment (nickroeker) [inline src/test/main.test.js]: Why is this being skipped now?

File: src/test/main.test.js [hunk 6]
```diff
@@ -376,41 +341,6 @@ describe('api', function() {
     });
 
 
-    it('public key validation', async () => {
```
Comment (nickroeker) [inline src/test/main.test.js]: Why was this removed? This seems especially important to the change.

File: src/test/main.test.js [hunk 7]
```diff
@@ -80,32 +87,6 @@ describe('Secret master key database creation', () => {
 
     });
 
-    it.skip('Fails with non-master key', async () => {
```
Comment (nickroeker) [inline src/test/main.test.js]: Why was this removed? This seems important to the change.

File: src/test/main.test.js (hunk 3 above)
Comment (MontyThibault) [reply src/test/main.test.js]: I think I'd prefer the full header/footer model so that we don't make any transformations of the key. It was like that with the old crypto setup because I had to manually interpret the base64 string. The new js crypto setup (and openssl) use the full key, so I don't see any reason to deviate from that structure.

https://bluzelle.slack.com/archives/G9LG77PEV/p1564511398032000

File: src/test/main.test.js (hunk 4 above)
Comment (MontyThibault) [reply src/test/main.test.js]: qa coverage

File: src/test/main.test.js (hunk 7 above)
Comment (MontyThibault) [reply src/test/main.test.js]: qa repo coverage

File: src/test/main.test.js (hunk 6 above)
Comment (MontyThibault) [reply src/test/main.test.js]: refactored into 'rejects invalid key' and 'rejects non-matching keys'

**Comment by MontyThibault** [comment]
- Refactored crypto into a validation layer and a signing layer.
- Reverted changes to key inputs.
- Testing changes.
- Added public/private key validation on startup.

File: src/swarmClient/03_crypto_verify_layer.js [hunk 8]
```diff
@@ -111,33 +81,39 @@ module.exports = class Crypto {
             }
         }   
 
-        if(!verify(Buffer.from(signed_bin), Buffer.from(bzn_envelope.getSignature()), bzn_envelope.getSender())) {            
-            this.log && this.log('Bluzelle: signature failed to verify: ' + Buffer.from(bin).toString('hex'));
-        }
 
-        this.onIncomingMsg(bzn_envelope);
+        const sender = format_public(bzn_envelope.getSender());
```
Comment (nickroeker) [inline src/swarmClient/03_crypto_verify_layer.js]: This does the format on every message if I understand correctly, which is not ideal in terms of performance.

File: src/swarmClient/07_crypto_sign_layer.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+        let sig;
+
+        try {
+
+            const s = crypto.createSign('sha256');
+
+            s.update(signed_bin);
+
+            sig = s.sign(format_private(this.private_pem));
```
Comment (nickroeker) [inline src/swarmClient/07_crypto_sign_layer.js]: This does the format on every message if I understand correctly, which is not ideal in terms of performance.

File: src/swarmClient/03_crypto_verify_layer.js (hunk 8 above)
Comment (MontyThibault) [reply src/swarmClient/03_crypto_verify_layer.js]: The format_public function is simple string manipulation and does not contain anything cryptographic.

File: src/swarmClient/07_crypto_sign_layer.js (hunk 9 above)
Comment (MontyThibault) [reply src/swarmClient/07_crypto_sign_layer.js]: See other comment.

File: src/swarmClient/main.js (hunk 1 above)
Comment (MontyThibault) [reply src/swarmClient/main.js]: It was resolved to load and verify keys at startup. See main.js:34 and main.test.js:376-411.

Comment (nickroeker) [reply src/swarmClient/main.js]: It was unclear that the conversation mentioned was in relation to this issue, my apologies. I stand by my original statement that we should not be regressing startup behaviour, especially in an unrelated commit/ticket/topic. This PR is only to address crypto efficiencies, and this is not a necessary change for that.

If we continue with the verification approach, we also need to change documentation and suggestions that, for example, you don't need a private key (or valid key pair at all) to perform non-write operations. Let's properly story this out and qualify it before implementing it.

File: src/swarmClient/03_crypto_verify_layer.js (hunk 8 above)
Comment (nickroeker) [reply src/swarmClient/03_crypto_verify_layer.js]: True, it's not a huge overhead in comparison, and it's only when crypto is being done which should mostly overshadow it. Microbenchmarking locally says the format operation is ~25% of the time crypto takes (for small messages), which is actually fairly significant but also a deeply flawed testing methodology.

Since this client is designed to operate in a serverless environment where time is money, and usually CPU speed isn't overly quick to begin with, we should avoid adding additional overhead between releases unless it is functionally required (e.g. no crypto -> crypto).

Anyhow, due to the likely low overall impact we can leave this for functional/performance qualification, but a note for future per-message changes.

File: src/swarmClient/main.js (hunk 1 above)
Comment (MontyThibault) [reply src/swarmClient/main.js]: > need to change documentation 

The existing documentation is consistent with this change. 

> ...suggestions that, for example, you don't need a private key (or valid key pair at all) to perform non-write operations

The agreed-upon method was for a valid key pair to be necessary to create a client object. 

> unrelated commit/ticket/topic

All crypto behavior is relevant to this ticket.

The startup needs to be re-implemented with the new crypto; that's why I needed to clarify the expected behavior.

File: src/swarmClient/03_crypto_verify_layer.js (hunk 8 above)
Comment (MontyThibault) [reply src/swarmClient/03_crypto_verify_layer.js]: A profiler would be the most appropriate approach to optimization when the time comes.
